File: Classes/QR-Code-Scanner/QRCodeScannerVC.m
//
//  QR-Code-Scanner-VC.m
//  Prey
//
//  Created by Javier Cala Uribe on 30/03/16.
//  Copyright © 2016 Fork Ltd. All rights reserved.
//

#import "QRCodeScannerVC.h"
#import "PreyDeployment.h"
#import "PreyRestHttp.h"
#import "Constants.h"

@interface QRCodeScannerVC ()

@property (nonatomic, strong) AVCaptureDevice *device;
@property (nonatomic, strong) AVCaptureDeviceInput *input;
@property (nonatomic, strong) AVCaptureSession *session;
@property (nonatomic, strong) AVCaptureMetadataOutput *output;
@property (nonatomic, strong) AVCaptureVideoPreviewLayer *preview;

@end

@implementation QRCodeScannerVC

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self.view setBackgroundColor:[UIColor redColor]];
    
    CGFloat widthScreen     = [[UIScreen mainScreen] bounds].size.width;
    UINavigationBar *navBar = [[UINavigationBar alloc] initWithFrame: CGRectMake(0, 0, widthScreen, 44)];
    navBar.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleBottomMargin;
    [self.view addSubview:navBar];
    
    UINavigationItem  *navItem = [[UINavigationItem alloc] initWithTitle:NSLocalizedString(@"Prey Control Panel", nil)];
    navItem.leftBarButtonItem  = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel
                                                                               target:self action:@selector(cancel:)];
    [navBar pushNavigationItem:navItem animated: NO];

    if ([self isCameraAvailable]) {
        // config session QR-Code
        [self setupScanner];
        
        // start scannning
        [self startScanning];        
    }
}

- (void)cancel:(id)sender {
    [self dismissViewControllerAnimated:YES completion:NULL];
}

- (BOOL)isCameraAvailable;
{
    NSArray *videoDevices = [AVCaptureDevice devicesWithMediaType:AVMediaTypeVideo];
    return [videoDevices count] > 0;
}

- (void)startScanning {
    [self.session startRunning];
}

- (void)stopScanning {
    [self.session stopRunning];
}

- (void) setupScanner {
    self.device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    
    self.input = [AVCaptureDeviceInput deviceInputWithDevice:self.device error:nil];
    
    self.session = [[AVCaptureSession alloc] init];
    
    self.output = [[AVCaptureMetadataOutput alloc] init];
    [self.session addOutput:self.output];
    [self.session addInput:self.input];
    
    [self.output setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
    self.output.metadataObjectTypes = @[AVMetadataObjectTypeQRCode];
    
    self.preview                    = [AVCaptureVideoPreviewLayer layerWithSession:self.session];
    self.preview.videoGravity       = AVLayerVideoGravityResizeAspectFill;
    self.preview.frame              = CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height);
    
    AVCaptureConnection *con        = self.preview.connection;
    con.videoOrientation            = AVCaptureVideoOrientationPortrait;
    
    [self.view.layer insertSublayer:self.preview atIndex:0];
    
    CGSize  screen      = [[UIScreen mainScreen] bounds].size;
    CGFloat widthLbl    = screen.width;
    CGFloat fontSize    = (IS_IPAD) ? 16.f : 12.f;
    NSString *message   = (IS_IPAD) ?   NSLocalizedString(@"Visit panel.preyproject.com/qr on your computer and scan the QR code",nil) :
                                        NSLocalizedString(@"Visit panel.preyproject.com/qr \non your computer and scan the QR code",nil);
    
    UILabel *infoQR         = [[UILabel alloc] initWithFrame:CGRectMake(0, screen.height-50, widthLbl, 50)];
    infoQR.textColor        = [UIColor colorWithRed:.3019f green:.3411f blue:.4f alpha:0.7f];
    infoQR.backgroundColor  = [UIColor whiteColor];
    infoQR.textAlignment    = NSTextAlignmentCenter;
    infoQR.font             = [UIFont fontWithName:@"OpenSans-Bold" size:fontSize];
    infoQR.text             = message;
    infoQR.numberOfLines    = 2;
    infoQR.adjustsFontSizeToFitWidth = YES;
    
    [self.view addSubview:infoQR];
    
    CGFloat qrZoneSize  = (IS_IPAD) ? screen.width*0.6f : screen.width*0.78f;
    CGFloat qrZonePosY  = (screen.height - qrZoneSize)/2;
    CGFloat qrZonePosX  = (screen.width  - qrZoneSize)/2;
    UIImageView *qrZone = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"qr-zone"]];
    qrZone.frame        = CGRectMake(qrZonePosX, qrZonePosY, qrZoneSize, qrZoneSize);
    
    [self.view addSubview:qrZone];
}

#pragma mark AVCaptureMetadataOutputObjectsDelegate

- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputMetadataObjects:(NSArray *)metadataObjects
       fromConnection:(AVCaptureConnection *)connection
{
    for(AVMetadataObject *current in metadataObjects)
    {
        if ([current isKindOfClass:[AVMetadataMachineReadableCodeObject class]])
        {
            if ([self respondsToSelector:@selector(scanViewController:didSuccessfullyScan:)])
            {
                NSString *scannedValue = [((AVMetadataMachineReadableCodeObject *) current) stringValue];
                [self scanViewController:self didSuccessfullyScan:scannedValue];
            }
        }
    }
}

- (void)scanViewController:(QRCodeScannerVC *)aCtler didSuccessfullyScan:(NSString *)aScannedValue {

    //NSLog(@"Code: %@", aScannedValue);
    NSString *validQr  = @"prey?api_key=";
    NSString *checkQr  = (aScannedValue.length > validQr.length) ? [aScannedValue substringToIndex:validQr.length]   : @"";
    NSString *apikeyQr = (aScannedValue.length > validQr.length) ? [aScannedValue substringFromIndex:validQr.length] : @"";
    
    [self stopScanning];

    [self dismissViewControllerAnimated:YES completion:^{
        
        if ([checkQr isEqualToString:validQr])
            [[PreyDeployment instance] addDeviceForApiKey:apikeyQr fromQRCode:YES];
        else
            [PreyRestHttp displayErrorAlert:NSLocalizedString(@"The scanned QR code is invalid", nil)
                                      title:NSLocalizedString(@"Couldn't add your device",nil)];
    }];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: Prey/Classes/QRCodeScannerVC.swift
//
//  QRCodeScannerVC.swift
//  Prey
//
//  Created by Javier Cala Uribe on 19/07/16.
//  Copyright © 2016 Prey, Inc. All rights reserved.
//

import AVFoundation
import Foundation
import UIKit

class QRCodeScannerVC: UIViewController, AVCaptureMetadataOutputObjectsDelegate {
    // MARK: Properties

    let device: AVCaptureDevice! = AVCaptureDevice.default(for: AVMediaType.video)
    let session: AVCaptureSession = .init()
    let output: AVCaptureMetadataOutput = .init()

    var preview: AVCaptureVideoPreviewLayer!

    // MARK: Init

    override func viewDidLoad() {
        super.viewDidLoad()

        // View title for GAnalytics
        // self.screenName = "QRCodeScanner"

        // Set background color
        view.backgroundColor = UIColor.black

        // Config navigationBar
        let widthScreen = UIScreen.main.bounds.size.width
        let navBar = UINavigationBar(frame: CGRect(x: 0, y: 0, width: widthScreen, height: 44))
        navBar.autoresizingMask = [.flexibleWidth, .flexibleBottomMargin]
        view.addSubview(navBar)

        // Config navItem
        let navItem = UINavigationItem(title: "")
        navItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancel))
        navBar.pushItem(navItem, animated: false)

        // Check camera available
        guard isCameraAvailable() else {
            displayErrorAlert("Couldn't add your device".localized,
                              titleMessage: "Error camera isn't available".localized)
            return
        }

        // Camera permission is requested here (and only here) so the user
        // sees the prompt on the QR scan screen instead of on app launch.
        requestCameraAccessIfNeeded { [weak self] granted in
            guard let self = self else { return }
            guard granted else {
                self.showCameraDeniedAlert()
                return
            }
            self.configureSession()
        }
    }

    /// Decision for the camera access flow, given an authorization status.
    /// Kept as a pure function so it can be unit-tested without touching AVFoundation.
    enum CameraAccessDecision: Equatable {
        case grant
        case prompt
        case deny
    }

    static func cameraAccessDecision(for status: AVAuthorizationStatus) -> CameraAccessDecision {
        switch status {
        case .authorized:
            return .grant
        case .notDetermined:
            return .prompt
        case .denied, .restricted:
            return .deny
        @unknown default:
            return .deny
        }
    }

    /// Request camera access if the status is notDetermined; otherwise resolve immediately.
    private func requestCameraAccessIfNeeded(completion: @escaping (Bool) -> Void) {
        switch QRCodeScannerVC.cameraAccessDecision(for: AVCaptureDevice.authorizationStatus(for: .video)) {
        case .grant:
            completion(true)
        case .prompt:
            AVCaptureDevice.requestAccess(for: .video) { granted in
                DispatchQueue.main.async { completion(granted) }
            }
        case .deny:
            completion(false)
        }
    }

    private func showCameraDeniedAlert() {
        let alert = UIAlertController(
            title: "Enable Camera".localized,
            message: "Prey uses the device's camera only to scan the QR code and add the device.".localized,
            preferredStyle: .alert
        )
        alert.addAction(UIAlertAction(title: "Go to Settings".localized, style: .default) { _ in
            if let url = URL(string: UIApplication.openSettingsURLString) {
                UIApplication.shared.open(url)
            }
            self.dismiss(animated: true)
        })
        alert.addAction(UIAlertAction(title: "Cancel".localized, style: .cancel) { _ in
            self.dismiss(animated: true)
        })
        present(alert, animated: true)
    }

    private func configureSession() {
        do {
            let inputDevice: AVCaptureDeviceInput = try AVCaptureDeviceInput(device: device)
            setupScanner(inputDevice)
            startScanning()
        } catch let error as NSError {
            PreyLogger("QrCode error: \(error.localizedDescription)")
            displayErrorAlert("Couldn't add your device".localized,
                              titleMessage: "The scanned QR code is invalid".localized)
        }
    }

    // MARK: Methods

    /// Setup scanner
    func setupScanner(_ input: AVCaptureDeviceInput) {
        // Config session
        session.addOutput(output)
        session.addInput(input)

        // Config output
        output.setMetadataObjectsDelegate(self, queue: DispatchQueue.main)
        output.metadataObjectTypes = [AVMetadataObject.ObjectType.qr]

        // Config preview
        preview = AVCaptureVideoPreviewLayer(session: session)
        preview.videoGravity = AVLayerVideoGravity.resizeAspectFill
        preview.frame = CGRect(x: 0, y: 0, width: view.frame.size.width, height: view.frame.size.height)
        preview.connection?.videoOrientation = .portrait
        view.layer.insertSublayer(preview, at: 0)

        // Config label
        let screen = UIScreen.main.bounds.size
        let widthLbl = screen.width
        let fontSize: CGFloat = IS_IPAD ? 16.0 : 12.0
        let message = IS_IPAD ? "Visit panel.preyproject.com/qr on your computer and scan the QR code".localized :
            "Visit panel.preyproject.com/qr \non your computer and scan the QR code".localized

        let infoQR = UILabel(frame: CGRect(x: 0, y: screen.height - 50, width: widthLbl, height: 50))
        infoQR.textColor = UIColor(red: 0.3019, green: 0.3411, blue: 0.4, alpha: 0.7)
        infoQR.backgroundColor = UIColor.white
        infoQR.textAlignment = .center
        infoQR.font = UIFont(name: fontTitilliumRegular, size: fontSize)
        infoQR.text = message
        infoQR.numberOfLines = 2
        infoQR.adjustsFontSizeToFitWidth = true
        view.addSubview(infoQR)

        // Config QrZone image
        let qrZoneSize = IS_IPAD ? screen.width * 0.6 : screen.width * 0.78
        let qrZonePosY = (screen.height - qrZoneSize) / 2
        let qrZonePosX = (screen.width - qrZoneSize) / 2
        let qrZoneImg = UIImageView(image: UIImage(named: "QrZone"))
        qrZoneImg.frame = CGRect(x: qrZonePosX, y: qrZonePosY, width: qrZoneSize, height: qrZoneSize)
        view.addSubview(qrZoneImg)
    }

    /// Success scan
    func successfullyScan(_ scannedValue: NSString) {
        let validQr = "prey?api_key=" as NSString
        let checkQr: NSString = (scannedValue.length > validQr.length) ? scannedValue.substring(to: validQr.length) as NSString : "" as NSString
        let apikeyQr: NSString = (scannedValue.length > validQr.length) ? scannedValue.substring(from: validQr.length) as NSString : "" as NSString

        stopScanning()

        dismiss(animated: true, completion: { () in
            if checkQr.isEqual(to: validQr as String) {
                PreyDeployment.sharedInstance.addDeviceWith(apikeyQr as String, fromQRCode: true)
            } else {
                displayErrorAlert("The scanned QR code is invalid".localized,
                                  titleMessage: "Couldn't add your device".localized)
            }
        })
    }

    func startScanning() {
        DispatchQueue.global(qos: .background).async {
            self.session.startRunning()
        }
    }

    func stopScanning() {
        session.stopRunning()
    }

    func isCameraAvailable() -> Bool {
        let discoverySession = AVCaptureDevice.DiscoverySession(deviceTypes: [.builtInWideAngleCamera], mediaType: .video, position: .unspecified)
        return !discoverySession.devices.isEmpty
    }

    @objc func cancel() {
        dismiss(animated: true, completion: nil)
    }

    // MARK: AVCaptureMetadataOutputObjectsDelegate

    /// CaptureOutput
    func metadataOutput(_: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from _: AVCaptureConnection) {
        for current in metadataObjects {
            if current is AVMetadataMachineReadableCodeObject {
                if let scannedValue = (current as! AVMetadataMachineReadableCodeObject).stringValue {
                    successfullyScan(scannedValue as NSString)
                }
            }
        }
    }
}
